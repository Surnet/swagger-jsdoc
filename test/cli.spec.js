const fs = require('fs');
const { promisify } = require('util');
const { exec } = require('child_process');

const sh = promisify(exec);
const dir = process.env.PWD;
const bin = `${dir}/bin/swagger-jsdoc.js`;

describe('command line interface', () => {
  it('help menu is default fallback when no arguments', async () => {
    const result = await sh(`${bin}`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('help menu works', async () => {
    const result = await sh(`${bin} -h`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should require a definition file', async () => {
    const result = await sh(`${bin} wrongDefinition`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should require an info object in the definition', async () => {
    const result = await sh(`${bin} -d test/files/v2/empty_definition.js`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should require title and version in the info object', async () => {
    const result = await sh(`${bin} -d test/files/v2/wrong_definition.js`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should require arguments with jsDoc data about an API', async () => {
    const result = await sh(`${bin} -d example/v2/swaggerDef.js`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should create swagger.json by default when the API input is good', async () => {
    const result = await sh(
      `${bin} -d example/v2/swaggerDef.js example/v2/routes.js`
    );
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('swagger.json');
    expect(specification.nlink).not.toBe(0);
  });

  it('should create swagger.json by default when the API input is from definition file', async () => {
    const result = await sh(`${bin} -d test/files/v2/api_definition.js`);
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('swagger.json');
    expect(specification.nlink).not.toBe(0);
  });

  it('should accept custom configuration for output specification', async () => {
    const result = await sh(
      `${bin} -d example/v2/swaggerDef.js -o customSpec.json example/v2/routes.js`
    );
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('customSpec.json');
    expect(specification.nlink).not.toBe(0);
  });

  it('should create a YAML swagger spec when a custom output configuration with a .yaml extension is used', async () => {
    const result = await sh(
      `${bin} -d example/v2/swaggerDef.js -o customSpec.yaml example/v2/routes.js`
    );
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('customSpec.yaml');
    expect(specification.nlink).not.toBe(0);
  });

  it('should allow a JavaScript definition file', async () => {
    const result = await sh(`${bin} -d test/files/v2/api_definition.js`);
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('swagger.json');
    expect(specification.nlink).not.toBe(0);
  });

  it('should allow a JSON definition file', async () => {
    const result = await sh(`${bin} -d test/files/v2/api_definition.json`);
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('swagger.json');
    expect(specification.nlink).not.toBe(0);
  });

  it('should allow a YAML definition file', async () => {
    const result = await sh(`${bin} -d test/files/v2/api_definition.yaml`);
    expect(result.stdout).toBe('Swagger specification is ready.\n');
    const specification = fs.statSync('swagger.json');
    expect(specification.nlink).not.toBe(0);
  });

  it('should reject definition file with invalid YAML syntax', async () => {
    const result = await sh(`${bin} -d test/files/v2/wrong_syntax.yaml`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should reject definition file with non-JSON compatible YAML syntax', async () => {
    const result = await sh(`${bin} -d test/files/v2/non_json_compatible.yaml`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should reject definition file with invalid JSON syntax', async () => {
    const result = await sh(`${bin} -d test/files/v2/wrong_syntax.json`);
    expect(result.stdout).toMatchSnapshot();
  });

  it('should reject bad YAML identation with feedback: upper line', async () => {
    await expect(
      sh(
        `${bin} -d example/v2/swaggerDef.js test/files/v2/wrong-yaml-identation1.js`
      )
    ).rejects.toThrow();
  });

  it('should reject bad YAML identation with feedback: same line', async () => {
    await expect(
      sh(
        `${bin} -d example/v2/swaggerDef.js test/files/v2/wrong-yaml-identation2.js`
      )
    ).rejects.toThrow();
  });

  afterAll(() => {
    fs.unlinkSync(`${dir}/swagger.json`);
    fs.unlinkSync(`${dir}/customSpec.json`);
    fs.unlinkSync(`${dir}/customSpec.yaml`);
    fs.unlinkSync(`${dir}/customSpec.yml`);
  });
});
