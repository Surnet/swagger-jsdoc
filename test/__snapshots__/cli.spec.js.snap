// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command line interface help menu is default fallback when no arguments 1`] = `
"Usage: swagger-jsdoc [options] <path ...>

Options:
  -V, --version                     output the version number
  -d, --definition <swaggerDef.js>  Input swagger definition.
  -o, --output [swaggerSpec.json]   Output swagger specification.
  -h, --help                        display help for command
"
`;

exports[`command line interface help menu works 1`] = `
"Usage: swagger-jsdoc [options] <path ...>

Options:
  -V, --version                     output the version number
  -d, --definition <swaggerDef.js>  Input swagger definition.
  -o, --output [swaggerSpec.json]   Output swagger specification.
  -h, --help                        display help for command
"
`;

exports[`command line interface should reject definition file with invalid JSON syntax 1`] = `
"Error while loading definition file 'test/files/v2/wrong_syntax.json':
Unexpected token t in JSON at position 18
"
`;

exports[`command line interface should reject definition file with invalid YAML syntax 1`] = `
"Error while loading definition file 'test/files/v2/wrong_syntax.yaml':
tag suffix cannot contain exclamation marks at line 2, column 5:
      !!!title: Hello World
        ^
"
`;

exports[`command line interface should reject definition file with non-JSON compatible YAML syntax 1`] = `
"Error while loading definition file 'test/files/v2/non_json_compatible.yaml':
unknown tag !<tag:yaml.org,2002:js/undefined> at line 3, column 3:
      version: 1.0.0
      ^
"
`;

exports[`command line interface should require a definition file 1`] = `
"Definition file is required.
You can do that, for example: 
$ swag-jsdoc -d swaggerDef.js wrongDefinition
Usage: swagger-jsdoc [options] <path ...>

Options:
  -V, --version                     output the version number
  -d, --definition <swaggerDef.js>  Input swagger definition.
  -o, --output [swaggerSpec.json]   Output swagger specification.
  -h, --help                        display help for command
"
`;

exports[`command line interface should require an info object in the definition 1`] = `
"Definition file should contain an info object!
More at http://swagger.io/specification/#infoObject
"
`;

exports[`command line interface should require arguments with jsDoc data about an API 1`] = `
"You must provide sources for reading API files.
Either add filenames as arguments, or add an \\"apis\\" key in your definitions file.
"
`;

exports[`command line interface should require title and version in the info object 1`] = `
"The title and version properties are required!
More at http://swagger.io/specification/#infoObject
"
`;
